src/sys/unix/net.rs

```diff
@@ -64,32 +64,76 @@ pub(crate) fn new_socket(
     socket
 }
 
-pub(crate) fn socket_addr(addr: &SocketAddr) -> (*const libc::sockaddr, libc::socklen_t) {
-    use std::mem::size_of_val;
+/// A type with the same memory layout as `libc::sockaddr`. Used in converting Rust level
+/// SocketAddr* types into their system representation. The benefit of this specific
+/// type over using `libc::sockaddr_storage` is that this type is exactly as large as it
+/// needs to be and not a lot larger. And it can be initialized cleaner from Rust.
+#[repr(C)]
+pub(crate) union SocketAddrCRepr {
+    v4: libc::sockaddr_in,
+    v6: libc::sockaddr_in6,
+}
+
+impl SocketAddrCRepr {
+    pub fn as_ptr(&self) -> *const libc::sockaddr {
+        self as *const _ as *const libc::sockaddr
```

Comment:
I'm not 100% on the rules when it comes to unions, I don't use them that often, so I have a question: does this require a `#[repr(transparent)]` attribute? Or is their a guarantee it contains no padding (at the start at least).

Reply:
`#[repr(transparent)]` does not work on unions. It also only works on structs or enums with a single field.

I think this is currently correctly implemented. A Rust `#[repr(C)]` union is supposed to have the same memory layout as a C enum. And a pointer to a C enum should always be a valid pointer to each field in it.